src/DynamicRouterBcLayer.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+/*
+ * This file is part of the Symfony CMF package.
+ *
+ * (c) Symfony CMF
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Cmf\Component\Routing;
+
+use Symfony\Cmf\Component\Routing\Event\Events;
+use Symfony\Cmf\Component\Routing\Event\RouterGenerateEvent;
+use Symfony\Component\EventDispatcher\LegacyEventDispatcherProxy;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+$urlGeneratorParameters = (new \ReflectionClass(UrlGeneratorInterface::class))->getMethod('generate')->getParameters();
+$urlGeneratorNameParameter = array_shift($urlGeneratorParameters);
+if ($urlGeneratorNameParameter && $urlGeneratorNameParameter->hasType() && 'string' === $urlGeneratorNameParameter->getType()) {
+    /**
+     * @internal
+     */
+    class DynamicRouterBcLayer extends DynamicRouterBaseBcLayer
+    {
+        public function generate(string $name, $parameters = [], $referenceType = UrlGeneratorInterface::ABSOLUTE_PATH)
+        {
+            return $this->doGenerate($name, $parameters, $referenceType);
+        }
+    }
+} else {
+    /**
+     * @internal
+     */
+    class DynamicRouterBcLayer extends DynamicRouterBaseBcLayer
+    {
+        public function generate($name, $parameters = [], $referenceType = UrlGeneratorInterface::ABSOLUTE_PATH)
+        {
+            if (!is_string($name)) {
+                @trigger_error(sprintf('Passing an object as the route name is deprecated in symfony-cmf/Routing v2.2 and will not work in Symfony 5.0. Pass the `RouteObjectInterface::OBJECT_BASED_ROUTE_NAME` constant as the route name and the object as "%s" parameter in the parameters array.', RouteObjectInterface::ROUTE_OBJECT), E_USER_DEPRECATED);
+
+                $parameters[RouteObjectInterface::ROUTE_OBJECT] = $name;
+                $name = RouteObjectInterface::OBJECT_BASED_ROUTE_NAME;
+            }
+
+            return $this->doGenerate($name, $parameters, $referenceType);
+        }
+    }
+}
+
+/**
+ * @internal
+ */
+abstract class DynamicRouterBaseBcLayer
+{
+    protected function doGenerate($name, $parameters, $referenceType)
+    {
+        if ($this->eventDispatcher) {
+            $routeParam = $name;
+            if (RouteObjectInterface::OBJECT_BASED_ROUTE_NAME === $name && array_key_exists(RouteObjectInterface::ROUTE_OBJECT, $parameters) && is_object($parameters[RouteObjectInterface::ROUTE_OBJECT])) {
+                $routeParam = $parameters[RouteObjectInterface::ROUTE_OBJECT];
+            }
+
+            $event = new RouterGenerateEvent($routeParam, $parameters, $referenceType);
+            $this->doDispatch(Events::PRE_DYNAMIC_GENERATE, $event);
+
+            $name = $event->getRoute();
```

Comment:
hm, together with the code above, this will end up putting the route object into the name, no? i think we are missing a test that would reveil the problem.

i think the intention is that the event can alter what to generate. we could forward the parameters into the event class as we did before and not read the route object from the parameters. it will be the job of the generator to do that.

additionally, we should adjust the event class https://github.com/symfony-cmf/Routing/blob/098ea1d035e2d2bda804e330fb99d39d1726123e/src/Event/RouterGenerateEvent.php#L64 setter, getter and constructor to say that the route should be a string and having objects will start failing in symfony 5.

Reply:
@dbu I've added deprecations in the event class and a test to verify it.

But I'm not sure if I understand what you mean with

> i think the intention is that the event can alter what to generate. we could forward the parameters into the event class as we did before and not read the route object from the parameters. it will be the job of the generator to do that.

Can you explain it a bit more? I think this is the last piece to finish this PR!

Reply:
sorry for the late reply, instead of lots of time for OS, corona lockdown made me have not much time at all...

the idea of the event is to have a place where users can change parameters or even the route that will be generated. i think we should create that event with $name and $parameters like before, and remove the `$routeParam = $parameters[RouteObjectInterface::ROUTE_OBJECT];` part. if we get there, we create the event with a route object and then the line this comment is attached to will put the object into `$name` which we then pass to the generate method, which violates the restriction that $name must be string.

imho we should dispatch the event with unmodified `$name` rather than put the route object into $name.